### DISCUSSION TITLE
Some tasmota devices don't reconnect after router restart

### DISCUSSION THREAD

**Discussion by jbornema** [arendst/Tasmota#16372] [Q&A]
Hi guys,

I'd like to share a work-a-round for my problem with Gosung SP111 and EP2 devices.
The problem was that almost everytime I restarted my router/repeater I had to check the state of my smart plugs and had to manually restart those devices, that didn't reconnect to Wifi (not even after some days!). I tried all options of tasmota and even disabled WMM (Wireless Multimedia) on my router. That improved it a little bit and I had less devices not coming back to network, bit it still happend.

After some hours trying to understand tasmota rules and what works in default builds, I came up with this rule that finally sovled the reconnection problems for me completely that I'd like to share:

```
rule1 
ON wifi#connected DO ruletimer1 0 ENDON 
ON wifi#disconnected DO ruletimer1 60 ENDON 
ON rules#timer=1 DO restart 1 ENDON
```

It simply restarts the tasmota after 60 Seconds without Wifi and all my reconnection problems after router/repeater reboots are now solved.

I hope this might help others, too.
Cheers - Jens.

**Comment by LoadedScissors** [comment]
Thanks!

**Reply by georgk111** [reply]
Will work only when rules are compiled into tasmota (like most precompiled firmware, and no scriting is activated)?  
My routers reboot once in 24 hours and here I noticed most Tasmotas regain WiFi but occasionally a few don't. I was thinking of a similar way to reboot sensors every 24 hours. but with your script I get the reboot automatically every 24hours. Will try it on 1 sensor ...
By the way: "setoption57 1" was not helpfull in those cases of wifi loss due to rebooted routers. Any idea why? In theory it should reconnect after 44 minutes??

**Reply by jbornema** [reply]
I only use standard firmware that has WebQuery and rule support enabled by default - I think "pinging" the router webpage, using WebQuery based on @tsmoneta 's suggesting would be the most stable way to reboot the tasmota device on wifi loss that only comes back after reboot.

**Comment by tsmoneta** [comment]
Hi- I had originally tried this, but ran into two scenarios where it didn't work:

1. Accidental 'wifi 0'. The rule didn't trigger apparently because of the manual (intentional/unintentional) state change. Had to power cycle.
2. Access point in a bad state. The modules were connected, but couldn't pass data until the access point was rebooted.

My solution that covered all the situations required a compile to add 'ping' support. Less convenient, but handles the additional situations:

```
rule1
on Time#Minute|5 do ping8 <router IP> endon
on Ping#<router IP>#Success==0 do restart 1 endon
```

This ensures not only a WiFi connection, but that the WiFi connection is actually capable of passing data. And it recovers from an accidental 'wifi 0' as well.

**Reply by jbornema** [reply]
Good catch on 'wifi 0' and the bad AP state. I suggest to use WebQuery for those that don't want to compile Tasmata with ping support as described in the [docs](https://tasmota.github.io/docs/Rules/#watchdog-for-wi-fi-router-or-modem).

Now, assuming your router has a web interface and reponds to http get with a login page or so and stops doing that, when it enters the "bad state" - this simple rule should do the same job without the need of ping8:
```
rule1
on Time#Minute|5 do backlog WebQuery http://<router IP> get endon
on WebQuery#Data$!Done do restart 1 endon
```

**Comment by tsmoneta** [comment]
There are some things to be aware of if using webquery.

My router transfers 870KB for its login page. I have 111 Tasmota devices currently, which means 96.5MB TCP (~1Gb) transfer for the check every 5 minutes, if the router doesn't throttle and toss the DDOS on it's login. Ping uses a 64B UDP packet, and the pings are paced at one second intervals, so 7.1KB per second, for 8 seconds.

The second issue is microwave interference. The webquery is a single large transfer, making it susceptible to transmission error. When the microwave is on, a few of the pings are lost, but most still make it through avoiding spurious reboots.

**Reply by georgk111** [reply]
I will test your script with ping8 on a plug occasionally loosing connectivity (it is at the edge of wifi covering), it normally comes up after rebooting.

**Reply by jbornema** [reply]
You're right, some login pages are pretty large, but only the html code would be downloaded, not any other linked resource like images or scripts etc. I just added /favicon.ico to the uri and tested that on my router it's only 961 Bytes - that would minimize the load significatly. But you're right - the ping option is really better, still hoping it enters the standard builds some time in the future 👍.

**Reply by georgk111** [reply]
PS: I have a few devices with scripts enabled - so no rules possible. But Scripts should be able to ping an IP and reboot the device on IP loss. I never tried programming scripts myself. Any good links to come over basic script programming?

**Reply by tsmoneta** [reply]
https://tasmota.github.io/docs/Scripting-Language/#script-sections

There's pretty good documentation and examples.

**Comment by tsmoneta** [comment]
It's worthwhile setting up a build environment for Tasmota in any case. I turn off a few of the drivers I don't need in addition to turning on ping. The net is a smaller module that avoids having to load the minimal firmware in most cases during upgrades.

**Comment by Andiweli** [comment]
Is this still not solved?
Have the same issue of „not reconnecting when router restarts“ with some of my Tasmota plugs. 

And isn't it needed to close the syntax with "Rule1 1" to activate the rule?

**Comment by fragtion** [comment]
The proposed workaround seems to work well. If the page served by your router is relatively large, you could try request a nonexistent html page, as a 404 error doesn't break the script.

I set mine up as follows:

`Rule1 on Time#Minute|5 do backlog WebQuery http://192.168.0.1/0 get endon on WebQuery#Data$!Done do Restart 1 endon`

And activated it with:

`Rule1 1`

**Comment by Arnie97** [comment]
Seems that `Wifi off / on` also works around the problem, and doesn't mess up with last boot time etc.
So my watchdog now reads

    On Time#Minute|2 Do
        Ping4 192.168.1.1
    EndOn
    On Ping#192.168.1.1#Success == 0 Do
        Backlog WiFi off; Delay 10; WiFi on
    EndOn

**Reply by ChemicalNRG** [reply]
Is any of this incorporated in the firmware? I also have 1 of 1 device not reconnecting, with firmware not more than 1 month old.

Maybe it is possible to do a restart with the button if that is available?

**Reply by sfromis** [reply]
The posted rules are by definition not part of the firmware, but something you can decide to work with - if you already have access. Note that what's posted is to restart the router in hope of this being of benefit. Chances would depend a lot on what exactly happens, there can be many causes. If you want to try if it helps simply restarting the router, you could try that manually. The posted rules were to automate that, but not much point doing such if it wouldn't help already.
